2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+
+When writing code, you spend a lot of time burrowed in the minutiae of specifics. You ply your craft at the mercy of unbending syntax rules &mdash; rules that will bring a million-line codebase to its knees over a single missing character. You write small functions that, given an exact input, produce an exact output. Every time.
+
+But naming demands that you wrest your mind from granular, concrete details and think more generally, more categorically.
+
+The linguist Sam Hayakawa, in his book [Language in Thought and Action](https://www.amazon.com/Language-Thought-Action-S-I-Hayakawa/dp/0156482401/), describes a model of language he calls the Abstraction Ladder. On this conceptual ladder, the bottommost rung holds the most concrete idea. The topmost rung is occupied by a decidedly abstract, intangible concept. In between, ideas get more abstract the higher up they are. The Abstraction Ladder is straightforward to understand, but studying it leads to broad and profound insights.
+
+<img src="images/abstraction-ladder.svg" width="100%" alt="The Abstraction Ladder" />
+
+Consider the diagram above. Molly is on the second rung. She is far from a static entity, she's a being in-process, constantly changing, composed of numerous subsystems that are represented on the bottom rung. But from a cognitive, linguistic perspective, Molly seems like a single entity. Indeed, associating a name with a living being is often the first utterance of a toddler learning to speak - "Mama."
+
+The next rung up the ladder, "Dogs" seems like a small step. But, consider the enormous mental processing it takes for a toddler to learn to ignore the differences between the family dog Molly, a small bull terrier, and the gigantic great dane that lives down the street. Learning to abstract from "Molly" to "Dogs" takes significant mental development, but once it's learned, it's automatic, processed by the brain without the perception of thought.
+
+The linguist is concerned with the relationship between words, abstractions, and meaning. So too is the programmer. Names are your words. Your names communicate meaning and understanding. Yes, you write comments as well, but comments often stand in to explain names that aren't quite right. A well-chosen name can obviate the need for an explanatory comment.
+
+## Naming functions
+
+Communication is most effective when it spans various levels of abstraction, when it travels up and down the ladder. Just stating that you should consider abstract ideas when naming things in your code is not enough. Providing abstract naming advice also requires concrete examples in order to be convincing. Concrete examples explain abstract ideas. In order to explain the color green, you might use examples like leaves in summertime, the bottom light of a traffic signal, or the color of an emerald.
```

Comment:
Instead of "Just stating . . . " how about this: "It's not enough to recommend that you consider abstract ideas when naming things in your code." (suggestion for removing the word "Just" which most of us overuse.)

Reply:
Addressed in e2195e047595a91cdbc86b740c6e85f0250d5d13